# OpenAI API

## Install the OpenAI library

This notebook has been automatically translated to make it accessible to more people, please let me know if you see any typos.

First of all, in order to use the OpenAI API, it is necessary to install the OpenAI library. To do this, run the following command

In [100]:
%pip install --upgrade openai

## Import the OpenAI library

Once the library is installed, we import it to be able to use it in our code.

In [11]:
import openai

## Obtain an API Key

In order to use the OpenAI API, it is necessary to obtain an API Key. To do this, go to the [OpenAI](https://openai.com/) page and register. Once registered, go to the [API Keys](https://platform.openai.com/api-keys) section, and create a new API Key.

![open ai api key](https://raw.githubusercontent.com/maximofn/alfred/main/gifs/openaix2.gif)

Once we have it, we tell the openai API which is our API Key.

In [12]:
api_key = "Pon aquí tu API key"

## We create our first chatbot

With the OpenAI API it is very easy to create a simple chatbot, to which we are going to pass a prompt, and it will give us a response

First of all we have to choose which model we are going to use, in my case I am going to use the `gpt-3.5-turbo-1106` model which is a good model for this post, since for what we are going to do we do not need to use the best model. OpenAI has a list with all their [models](https://platform.openai.com/docs/models) and a page with the [prices](https://openai.com/pricing)

In [42]:
model = "gpt-3.5-turbo-1106"

Now we have to create a client that will communicate with the OpenAI API.

In [3]:
client = openai.OpenAI(api_key=api_key, organization=None)

As we can see we have passed our API Key. It is also possible to pass the organization, but in our case it is not necessary.

We create the prompt

In [44]:
promtp = "Cuál es el mejor lenguaje de programación para aprender?"

And now we can ask OpenAI for an answer.

In [45]:
response = client.chat.completions.create(
  model=model,
  messages=[{"role": "user", "content": f"{promtp}"}],
)

Let's see what the answer looks like

In [46]:
type(response), response

(openai.types.chat.chat_completion.ChatCompletion,
 ChatCompletion(id='chatcmpl-8RaHCm9KalLxj2PPbLh6f8A4djG8Y', choices=[Choice(finish_reason='stop', index=0, message=ChatCompletionMessage(content='No hay un "mejor" lenguaje de programación para aprender, ya que depende de tus intereses, objetivos y el tipo de desarrollo que te interese. Algunos lenguajes populares para empezar a aprender a programar incluyen Python, JavaScript, Java, C# y Ruby. Estos lenguajes son conocidos por su sintaxis clara y su versatilidad, lo que los hace buenos candidatos para principiantes. También es útil investigar qué lenguajes son populares en la industria en la que te gustaría trabajar, ya que el conocimiento de un lenguaje en demanda puede abrirte más oportunidades laborales. En resumen, la elección del lenguaje de programación para aprender dependerá de tus preferencias personales y de tus metas profesionales.', role='assistant', function_call=None, tool_calls=None))], created=1701584994, model='gpt-3

In [48]:
print(f"response.id = {response.id}")
print(f"response.choices = {response.choices}")
for i in range(len(response.choices)):
    print(f"response.choices[{i}] = {response.choices[i]}")
    print(f"\tresponse.choices[{i}].finish_reason = {response.choices[i].finish_reason}")
    print(f"\tresponse.choices[{i}].index = {response.choices[i].index}")
    print(f"\tresponse.choices[{i}].message = {response.choices[i].message}")
    content = response.choices[i].message.content.replace('\n\n', '\n\t\t')
    print(f"\t\tresponse.choices[{i}].message.content = \n\t\t{content}")
    print(f"\t\tresponse.choices[{i}].message.role = {response.choices[i].message.role}")
    print(f"\t\tresponse.choices[{i}].message.function_call = {response.choices[i].message.function_call}")
    print(f"\t\tresponse.choices[{i}].message.tool_calls = {response.choices[i].message.tool_calls}")
print(f"response.created = {response.created}")
print(f"response.model = {response.model}")
print(f"response.object = {response.object}")
print(f"response.system_fingerprint = {response.system_fingerprint}")
print(f"response.usage = {response.usage}")
print(f"\tresponse.usage.completion_tokens = {response.usage.completion_tokens}")
print(f"\tresponse.usage.prompt_tokens = {response.usage.prompt_tokens}")
print(f"\tresponse.usage.total_tokens = {response.usage.total_tokens}")

response.id = chatcmpl-8RaHCm9KalLxj2PPbLh6f8A4djG8Y
response.choices = [Choice(finish_reason='stop', index=0, message=ChatCompletionMessage(content='No hay un "mejor" lenguaje de programación para aprender, ya que depende de tus intereses, objetivos y el tipo de desarrollo que te interese. Algunos lenguajes populares para empezar a aprender a programar incluyen Python, JavaScript, Java, C# y Ruby. Estos lenguajes son conocidos por su sintaxis clara y su versatilidad, lo que los hace buenos candidatos para principiantes. También es útil investigar qué lenguajes son populares en la industria en la que te gustaría trabajar, ya que el conocimiento de un lenguaje en demanda puede abrirte más oportunidades laborales. En resumen, la elección del lenguaje de programación para aprender dependerá de tus preferencias personales y de tus metas profesionales.', role='assistant', function_call=None, tool_calls=None))]
response.choices[0] = Choice(finish_reason='stop', index=0, message=ChatCompletio

As we can see, it returns a lot of information.

For example `response.choices[0].finish_reason = stop` means that the model has stopped generating text because it has reached the end of the prompt. This comes in handy for debugging, as the possible values are `stop` which means the API returned the complete message, `length` which means the model output was incomplete because it was longer than `max_tokens` or model token limit, `function_call` the model decided to call a function, `content_filter` which means the content was skipped because of an OpenAI content limitation and `null` which means the API response was incomplete.

It also gives us token information to keep track of the money spent.

### Parameters

When asking OpenAI for an answer, we can pass a series of parameters to it so that it returns an answer more in accordance with what we want. Let's see which are the parameters that we can pass to it

 * Messages: List of messages that have been sent to the chatbot.
 * `model`: Model we want to use
 * `frequency_penalty`: Frequency penalty. The higher the value, the less likely the model will repeat the same response.
 * `max_tokens`: Maximum number of tokens the model can return.
 * `n`: Number of responses that we want the model to return.
 * `presence_penalty`: Presence penalty. The higher the value, the less likely the model will repeat the same response.
 * `seed`: Seed for text generation
 * `stop`: List of tokens indicating that the model should stop generating text.
 * If `stream`: If `True` the API will return a response each time the model generates a token. If `False` the API will return a response when the model has generated all tokens.
 * `temperature`: The higher the value, the more creative the model will be.
 * `top_p`: The higher the value, the more creative the model will be.
 * `user`: ID of the user who is talking to the chatbot
 * `timeout`: Maximum time we want to wait for the API to return a response.

Let's take a look at some of them

#### Messages

We can pass to the API a list of messages that have been sent to the chatbot. This is useful to pass the conversation history to the chatbot, so that it can generate a response more in line with the conversation. And to condition the chatbot's response to what it has been told previously.

We can also pass a system message to tell it how to behave.

##### Conversation history

Let's see an example of a conversation history.

In [112]:
promtp = "Hola, soy MaximoFN, ¿Cómo estás?"

response = client.chat.completions.create(
  model=model,
  messages=[{"role": "user", "content": f"{promtp}"}],
)
content = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content)

Hola MaximoFN, soy un modelo de inteligencia artificial diseñado para conversar y ayudar en lo que necesites. ¿En qué puedo ayudarte hoy?


He answered that he has no feelings and how can he help us. So if I ask him what my name is now, he won't know how to answer me.

In [113]:
promtp = "¿Me puedes decir cómo me llamo?"

response = client.chat.completions.create(
  model=model,
  messages=[{"role": "user", "content": f"{promtp}"}],
)
content = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content)

Lo siento, no tengo esa información. Pero puedes decírmelo tú.


To solve this, we pass the conversation history to you

In [117]:
promtp = "¿Me puedes decir cómo me llamo?"

response = client.chat.completions.create(
  model=model,
  messages=[
      {"role": "user", "content": "Hola, soy MaximoFN, ¿Cómo estás?"},
      {"role": "assistant", "content": "Hola MaximoFN, soy un modelo de inteligencia artificial diseñado para conversar y ayudar en lo que necesites. ¿En qué puedo ayudarte hoy?"},
      {"role": "user", "content": f"{promtp}"},
  ],
)
content = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content)

Tu nombre es MaximoFN.


##### Conditioning by examples

Now let's look at an example of how to condition the chatbot's response to what it has been told above. Now we ask it how to get the list of files in a directory in the terminal

In [121]:
promtp = "¿Cómo puedo listar los archivos de un directorio en la terminal?"

response = client.chat.completions.create(
  model=model,
  messages=[{"role": "user", "content": f"{promtp}"}],
)
content = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content)

En la terminal de un sistema operativo Unix o Linux, puedes listar los archivos de un directorio utilizando el comando `ls`. Por ejemplo, si quieres listar los archivos del directorio actual, simplemente escribe `ls` y presiona Enter. Si deseas listar los archivos de un directorio específico, puedes proporcionar la ruta del directorio después del comando `ls`, por ejemplo `ls /ruta/del/directorio`. Si deseas ver más detalles sobre los archivos, puedes usar la opción `-l` para obtener una lista detallada o `-a` para mostrar también los archivos ocultos.


If we now condition him with examples of short answers, let's see what he will answer us

In [124]:
promtp = "¿Cómo puedo listar los archivos de un directorio en la terminal?"

response = client.chat.completions.create(
  model=model,
  messages=[
      {"role": "user", "content": "Obtener las 10 primeras líneas de un archivo"},
      {"role": "assistant", "content": "head -n 10"},
      {"role": "user", "content": "Encontrar todos los archivos con extensión .txt"},
      {"role": "assistant", "content": "find . -name '*.txt"},
      {"role": "user", "content": "Dividir un archivo en varias páginas"},
      {"role": "assistant", "content": "split -l 1000"},
      {"role": "user", "content": "Buscar la dirección IP 12.34.56.78"},
      {"role": "assistant", "content": "nslookup 12.34.56.78"},
      {"role": "user", "content": "Obtener las 5 últimas líneas de foo.txt"},
      {"role": "assistant", "content": "tail -n 5 foo.txt"},
      {"role": "user", "content": "Convertir ejemplo.png en JPEG"},
      {"role": "assistant", "content": "convert example.png example.jpg"},
      {"role": "user", "content": "Create a git branch named 'new-feature"},
      {"role": "assistant", "content": "git branch new-feature"},
      {"role": "user", "content": f"{promtp}"},
  ],
)
content = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content)

Puede usar el comando `ls` en la terminal para listar los archivos de un directorio. Por ejemplo:
```
ls
```
Muestra los archivos y directorios en el directorio actual.


We have managed to give a shorter response

##### Conditioning with system message

We can pass you a system message to tell you how to behave.

In [126]:
promtp = "¿Cómo puedo listar los archivos de un directorio en la terminal?"

response = client.chat.completions.create(
  model=model,
  messages=[
      {"role": "system", "content": "Eres un experto asistente de terminal de ubuntu que responde solo con comandos de terminal"},
      {"role": "user", "content": f"{promtp}"},
  ],
)
content = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content)

Puedes listar los archivos de un directorio en la terminal usando el comando `ls`. Por ejemplo, para listar los archivos del directorio actual, simplemente escribe `ls` y presiona Enter. Si quieres listar los archivos de un directorio específico, puedes utilizar `ls` seguido de la ruta del directorio. Por ejemplo, `ls /ruta/del/directorio`.


In [128]:
promtp = "¿Cómo puedo listar los archivos de un directorio en la terminal?"

response = client.chat.completions.create(
  model=model,
  messages=[
      {"role": "system", "content": "Eres un experto asistente de terminal de ubuntu que responde solo con comandos de terminal"},
      {"role": "user", "content": "Obtener las 10 primeras líneas de un archivo"},
      {"role": "assistant", "content": "head -n 10"},
      {"role": "user", "content": "Encontrar todos los archivos con extensión .txt"},
      {"role": "assistant", "content": "find . -name '*.txt"},
      {"role": "user", "content": "Dividir un archivo en varias páginas"},
      {"role": "assistant", "content": "split -l 1000"},
      {"role": "user", "content": "Buscar la dirección IP 12.34.56.78"},
      {"role": "assistant", "content": "nslookup 12.34.56.78"},
      {"role": "user", "content": "Obtener las 5 últimas líneas de foo.txt"},
      {"role": "assistant", "content": "tail -n 5 foo.txt"},
      {"role": "user", "content": "Convertir ejemplo.png en JPEG"},
      {"role": "assistant", "content": "convert example.png example.jpg"},
      {"role": "user", "content": "Create a git branch named 'new-feature"},
      {"role": "assistant", "content": "git branch new-feature"},
      {"role": "user", "content": f"{promtp}"},
  ],
)
content = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content)

Puedes listar los archivos de un directorio en la terminal utilizando el comando "ls". Por ejemplo, para listar los archivos en el directorio actual, puedes ejecutar el comando "ls". Si deseas listar los archivos de otro directorio, simplemente especifica el directorio después del comando "ls", por ejemplo "ls /ruta/al/directorio".


#### Maximum number of response tokens

We can limit the number of tokens that the model can return. This is useful so that the model does not overshoot the response we want.

In [136]:
promtp = "¿Cuál es el mejor lenguaje de programación para aprender?"

response = client.chat.completions.create(
  model = model,
  messages = [{"role": "user", "content": f"{promtp}"}],
  max_tokens = 50,
)

content = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content)
print(f"\nresponse.choices[{i}].finish_reason = {response.choices[i].finish_reason}")

La respuesta a esta pregunta puede variar dependiendo de los intereses y objetivos individuales, ya que cada lenguaje de programación tiene sus propias ventajas y desventajas. Sin embargo, algunos de los lenguajes más

response.choices[0].finish_reason = length


As we can see the answer is half cut because it would exceed the token limit. Also now the stop reason is `length` instead of `stop`.

#### Creativity of the model through temperature

We can make the model more creative by the temperature. The higher the value, the more creative the model will be.

In [152]:
promtp = "¿Cuál es el mejor lenguaje de programación para aprender?"
temperature = 0

response = client.chat.completions.create(
  model = model,
  messages = [{"role": "user", "content": f"{promtp}"}],
  temperature = temperature,
)

content_0 = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content_0)

No hay un "mejor" lenguaje de programación para aprender, ya que la elección depende de los intereses y objetivos individuales. Algunos lenguajes populares para principiantes incluyen Python, JavaScript, Java y C#. Cada uno tiene sus propias ventajas y desventajas, por lo que es importante investigar y considerar qué tipo de desarrollo de software te interesa antes de elegir un lenguaje para aprender.


In [153]:
promtp = "¿Cuál es el mejor lenguaje de programación para aprender?"
temperature = 1

response = client.chat.completions.create(
  model = model,
  messages = [{"role": "user", "content": f"{promtp}"}],
  temperature = temperature,
)

content_1 = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content_1)

No hay un "mejor" lenguaje de programación para aprender, ya que la elección depende de los objetivos y preferencias individuales del programador. Sin embargo, algunos lenguajes populares para principiantes incluyen Python, JavaScript, Java y C++. Estos lenguajes son relativamente fáciles de aprender y tienen una amplia gama de aplicaciones en la industria de la tecnología. Es importante considerar qué tipo de proyectos o campos de interés te gustaría explorar al momento de elegir un lenguaje de programación para aprender.


In [154]:
print(content_0)
print(content_1)

No hay un "mejor" lenguaje de programación para aprender, ya que la elección depende de los intereses y objetivos individuales. Algunos lenguajes populares para principiantes incluyen Python, JavaScript, Java y C#. Cada uno tiene sus propias ventajas y desventajas, por lo que es importante investigar y considerar qué tipo de desarrollo de software te interesa antes de elegir un lenguaje para aprender.
No hay un "mejor" lenguaje de programación para aprender, ya que la elección depende de los objetivos y preferencias individuales del programador. Sin embargo, algunos lenguajes populares para principiantes incluyen Python, JavaScript, Java y C++. Estos lenguajes son relativamente fáciles de aprender y tienen una amplia gama de aplicaciones en la industria de la tecnología. Es importante considerar qué tipo de proyectos o campos de interés te gustaría explorar al momento de elegir un lenguaje de programación para aprender.


#### Creativity of the model using the top_p

We can make the model more creative by using the `top_p` parameter. The higher the value, the more creative the model will be.

In [155]:
promtp = "¿Cuál es el mejor lenguaje de programación para aprender?"
top_p = 0

response = client.chat.completions.create(
  model = model,
  messages = [{"role": "user", "content": f"{promtp}"}],
  top_p = top_p,
)

content_0 = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content_0)

No hay un "mejor" lenguaje de programación para aprender, ya que la elección depende de los intereses y objetivos individuales. Algunos lenguajes populares para principiantes incluyen Python, JavaScript, Java y C#. Cada uno tiene sus propias ventajas y desventajas, por lo que es importante investigar y considerar qué tipo de desarrollo de software te interesa antes de elegir un lenguaje para aprender.


In [156]:
promtp = "¿Cuál es el mejor lenguaje de programación para aprender?"
top_p = 1

response = client.chat.completions.create(
  model = model,
  messages = [{"role": "user", "content": f"{promtp}"}],
  top_p = top_p,
)

content_1 = response.choices[0].message.content.replace('\n\n', '\n\t\t')
print(content_1)

El mejor lenguaje de programación para aprender depende de los objetivos del aprendizaje y del tipo de programación que se quiera realizar. Algunos lenguajes de programación populares para principiantes incluyen Python, Java, JavaScript y Ruby. Sin embargo, cada lenguaje tiene sus propias ventajas y desventajas, por lo que es importante considerar qué tipo de proyectos o aplicaciones se quieren desarrollar antes de elegir un lenguaje de programación para aprender. Python es a menudo recomendado por su facilidad de uso y versatilidad, mientras que JavaScript es ideal para la programación web.


In [157]:
print(content_0)
print(content_1)

No hay un "mejor" lenguaje de programación para aprender, ya que la elección depende de los intereses y objetivos individuales. Algunos lenguajes populares para principiantes incluyen Python, JavaScript, Java y C#. Cada uno tiene sus propias ventajas y desventajas, por lo que es importante investigar y considerar qué tipo de desarrollo de software te interesa antes de elegir un lenguaje para aprender.
El mejor lenguaje de programación para aprender depende de los objetivos del aprendizaje y del tipo de programación que se quiera realizar. Algunos lenguajes de programación populares para principiantes incluyen Python, Java, JavaScript y Ruby. Sin embargo, cada lenguaje tiene sus propias ventajas y desventajas, por lo que es importante considerar qué tipo de proyectos o aplicaciones se quieren desarrollar antes de elegir un lenguaje de programación para aprender. Python es a menudo recomendado por su facilidad de uso y versatilidad, mientras que JavaScript es ideal para la programación w

#### Number of responses

We can ask the API to return more than one response. This is useful for the model to return several answers and so we can choose the one we like the most, for this we will set the parameters `temperature` and `top_p` to 1 to make the model more creative.

In [151]:
promtp = "¿Cuál es el mejor lenguaje de programación para aprender?"
temperature = 1
top_p = 1

response = client.chat.completions.create(
  model = model,
  messages = [{"role": "user", "content": f"{promtp}"}],
  temperature = temperature,
  top_p = top_p,
  n = 4
)

content_0 = response.choices[0].message.content.replace('\n\n', '\n\t\t')
content_1 = response.choices[1].message.content.replace('\n\n', '\n\t\t')
content_2 = response.choices[2].message.content.replace('\n\n', '\n\t\t')
content_3 = response.choices[3].message.content.replace('\n\n', '\n\t\t')
print(content_0)
print(content_1)
print(content_2)
print(content_3)

El mejor lenguaje de programación para aprender depende de tus objetivos y del tipo de aplicaciones que te interese desarrollar. Algunos de los lenguajes más populares para aprender son:
		1. Python: Es un lenguaje de programación versátil, fácil de aprender y con una amplia comunidad de desarrolladores. Es ideal para principiantes y se utiliza en una gran variedad de aplicaciones, desde desarrollo web hasta inteligencia artificial.
		2. JavaScript: Es el lenguaje de programación más utilizado en el desarrollo web. Es imprescindible para aquellos que quieren trabajar en el ámbito del desarrollo frontend y backend.
		3. Java: Es un lenguaje de programación muy popular en el ámbito empresarial, por lo que aprender Java puede abrirte muchas puertas laborales. Además, es un lenguaje estructurado que te enseñará conceptos importantes de la programación orientada a objetos.
		4. C#: Es un lenguaje de programación desarrollado por Microsoft que se utiliza especialmente en el desarrollo de apl

## Re-train OpenAI model

OpenAI offers the possibility to retrain its API models to obtain better results on our own data. This has the following advantages

 + Higher quality results are obtained for our data.
 + In a prompt we can give it examples to behave as we want, but only a few. This way, by retraining it we can give it many more.
 + Saving tokens due to shorter prompts. Since we have already trained it for our use case, we can give it fewer prompts to solve our tasks.
 + Lower latency requests. By calling our own models we will have less latency

### Data preparation

The OpenAI API asks us to provide the data in a `jsonl` file in the following format

``` json
{
    "messages":
    [
        {
            "role": "system", "content": "Marv is a factual chatbot that is also sarcastic."
        },
        {
            "role": "user", "content": "What's the capital of France?"
        },
        {
            "role": "assistant", "content": "Paris, as if everyone doesn't know that already."
        }
    
}
{
    "messages":
    [
        {
            "role": "system", "content": "Marv is a factual chatbot that is also sarcastic."
        },
        {
            "role": "user", "content": "Who wrote 'Romeo and Juliet'?"
        },
        {
            "role": "assistant", "content": "Oh, just some guy named William Shakespeare. Ever heard of him?"
        }
    
}
{
    "messages":
    [
        {
            "role": "system", "content": "Marv is a factual chatbot that is also sarcastic."
        },
        {
            "role": "user", "content": "How far is the Moon from Earth?"
        },
        {
            "role": "assistant", "content": "Around 384,400 kilometers. Give or take a few, like that really matters."
        }
    
}
```

With a maximum of 4096 tokens

#### Data validation

To save me some work, I have been passing one by one all my posts to chatgpt and I have told it to generate 10 `FAQ`s for each one in `CSV` format, because I doubted if it was going to be able to generate a format like the one requested in the `jsonl`. And it has generated a `CSV` with the following format for each post

``` csv
prompt,completion
What does Introduction to Python cover in the material provided, "Introduction to Python covers topics such as data types, operators, use of functions and classes, handling iterable objects, and use of modules. [Learn more](https://maximofn.com/python/)"
What are the basic data types in Python, "Python has 7 basic data types: text (`str`), numeric (`int`, `float`, `complex`), sequences (`list`, `tuple`, `range`), mapping (`dict`), sets (`set`, `frozenset`), boolean (`bool`) and binary (`bytes`, `bytearray`, `memoryview`). [More information](https://maximofn.com/python/)"
What are operators in Python and how are they used, "Operators in Python are special symbols used to perform operations such as addition, subtraction, multiplication and division between variables and values. They also include logical operators for comparisons. [Learn more](https://maximofn.com/python/)"
How is a function defined and used in Python, "In Python, a function is defined using the `def` keyword, followed by the function name and parentheses. Functions can have parameters and return values. They are used to encapsulate logic that can be reused throughout the code. [More information](https://maximofn.com/python/)"
What are Python classes and how are they used, "Python classes are the basis of object-oriented programming. They allow you to create objects that encapsulate data and functionality. Classes are defined using the `class` keyword, followed by the class name. [More information](https://maximofn.com/python/)"
...
```

Each `CSV` has 10 `FAQ`s

I am going to make a code that takes each `CSV` and generates two new `jsonl`s, one for training and one for validation.

In [59]:
import os

CSVs_path = "openai/faqs_posts"
percetn_train = 0.8
percetn_validation = 0.2

jsonl_train = os.path.join(CSVs_path, "train.jsonl")
jsonl_validation = os.path.join(CSVs_path, "validation.jsonl")

# Create the train.jsonl and validation.jsonl files
with open(jsonl_train, 'w') as f:
    f.write('')
with open(jsonl_validation, 'w') as f:
    f.write('')

for file in os.listdir(CSVs_path):  # Get all files in the directory
    if file.endswith(".csv"):    # Check if file is a csv
        csv = os.path.join(CSVs_path, file) # Get the path to the csv file
        number_of_lines = 0
        csv_content = []
        for line in open(csv, 'r'): # Read all lines in the csv file
            if line.startswith('prompt'):   # Skip the first line
                continue
            number_of_lines += 1    # Count the number of lines
            csv_content.append(line)    # Add the line to the csv_content list

        number_of_train = int(number_of_lines * percetn_train)  # Calculate the number of lines for the train.jsonl file
        number_of_validation = int(number_of_lines * percetn_validation)    # Calculate the number of lines for the validation.sjonl file

        for i in range(number_of_lines):
            prompt = csv_content[i].split(',')[0]
            response = ','.join(csv_content[i].split(',')[1:]).replace('\n', '').replace('"', '')
            if i > 0 and i <= number_of_train:
                # add line to train.jsonl
                with open(jsonl_train, 'a') as f:
                    f.write(f'{"{"}"messages": [{"{"}"role": "system", "content": "Eres un amable asistente dispuesto a responder."{"}"}, {"{"}"role": "user", "content": "{prompt}"{"}"}, {"{"}"role": "assistant", "content": "{response}"{"}"}]{"}"}\n')
            elif i > number_of_train and i <= number_of_train + number_of_validation:
                # add line to validation.csv
                with open(jsonl_validation, 'a') as f:
                    f.write(f'{"{"}"messages": [{"{"}"role": "system", "content": "Eres un amable asistente dispuesto a responder."{"}"}, {"{"}"role": "user", "content": "{prompt}"{"}"}, {"{"}"role": "assistant", "content": "{response}"{"}"}]{"}"}\n')

Once I have the two `jsonl`s, I run a [code](https://cookbook.openai.com/examples/chat_finetuning_data_prep) provided by OpenAI to check the `jsonl`s

First we validate the training ones

In [9]:
from collections import defaultdict
import json

# Format error checks
format_errors = defaultdict(int)

with open(jsonl_train, 'r', encoding='utf-8') as f:
    dataset = [json.loads(line) for line in f]

for ex in dataset:
    if not isinstance(ex, dict):
        format_errors["data_type"] += 1
        continue
        
    messages = ex.get("messages", None)
    if not messages:
        format_errors["missing_messages_list"] += 1
        continue
        
    for message in messages:
        if "role" not in message or "content" not in message:
            format_errors["message_missing_key"] += 1
        
        if any(k not in ("role", "content", "name", "function_call") for k in message):
            format_errors["message_unrecognized_key"] += 1
        
        if message.get("role", None) not in ("system", "user", "assistant", "function"):
            format_errors["unrecognized_role"] += 1
            
        content = message.get("content", None)
        function_call = message.get("function_call", None)
        
        if (not content and not function_call) or not isinstance(content, str):
            format_errors["missing_content"] += 1
    
    if not any(message.get("role", None) == "assistant" for message in messages):
        format_errors["example_missing_assistant_message"] += 1

if format_errors:
    print("Found errors:")
    for k, v in format_errors.items():
        print(f"{k}: {v}")
else:
    print("No errors found")

No errors found


And now those of validation

In [10]:
# Format error checks
format_errors = defaultdict(int)

with open(jsonl_validation, 'r', encoding='utf-8') as f:
    dataset = [json.loads(line) for line in f]

for ex in dataset:
    if not isinstance(ex, dict):
        format_errors["data_type"] += 1
        continue
        
    messages = ex.get("messages", None)
    if not messages:
        format_errors["missing_messages_list"] += 1
        continue
        
    for message in messages:
        if "role" not in message or "content" not in message:
            format_errors["message_missing_key"] += 1
        
        if any(k not in ("role", "content", "name", "function_call") for k in message):
            format_errors["message_unrecognized_key"] += 1
        
        if message.get("role", None) not in ("system", "user", "assistant", "function"):
            format_errors["unrecognized_role"] += 1
            
        content = message.get("content", None)
        function_call = message.get("function_call", None)
        
        if (not content and not function_call) or not isinstance(content, str):
            format_errors["missing_content"] += 1
    
    if not any(message.get("role", None) == "assistant" for message in messages):
        format_errors["example_missing_assistant_message"] += 1

if format_errors:
    print("Found errors:")
    for k, v in format_errors.items():
        print(f"{k}: {v}")
else:
    print("No errors found")

No errors found


#### Calculation of tokens

The maximum number of tokens for each example has to be 4096, so if we have longer examples only the first 4096 tokens will be used. So let's count the number of tokens that each `jsonl` has to know how much it will cost us to retrain the model.

But first we must install the `tiktoken` library, which is the tokenizer used by OpenAI and which will also help us to know how many tokens each `CSV` has, and therefore, how much it will cost us to retrain the model.

To install it we execute the following command

``` bash
pip install tiktoken
```

We create a few necessary functions

In [21]:
import tiktoken
import numpy as np

encoding = tiktoken.get_encoding("cl100k_base")

def num_tokens_from_messages(messages, tokens_per_message=3, tokens_per_name=1):
    num_tokens = 0
    for message in messages:
        num_tokens += tokens_per_message
        for key, value in message.items():
            num_tokens += len(encoding.encode(value))
            if key == "name":
                num_tokens += tokens_per_name
    num_tokens += 3
    return num_tokens

def num_assistant_tokens_from_messages(messages):
    num_tokens = 0
    for message in messages:
        if message["role"] == "assistant":
            num_tokens += len(encoding.encode(message["content"]))
    return num_tokens

def print_distribution(values, name):
    print(f"\n#### Distribution of {name}:")
    print(f"min:{min(values)}, max: {max(values)}")
    print(f"mean: {np.mean(values)}, median: {np.median(values)}")
    print(f"p5: {np.quantile(values, 0.1)}, p95: {np.quantile(values, 0.9)}")

In [22]:
# Warnings and tokens counts
n_missing_system = 0
n_missing_user = 0
n_messages = []
convo_lens = []
assistant_message_lens = []

with open(jsonl_train, 'r', encoding='utf-8') as f:
    dataset = [json.loads(line) for line in f]

for ex in dataset:
    messages = ex["messages"]
    if not any(message["role"] == "system" for message in messages):
        n_missing_system += 1
    if not any(message["role"] == "user" for message in messages):
        n_missing_user += 1
    n_messages.append(len(messages))
    convo_lens.append(num_tokens_from_messages(messages))
    assistant_message_lens.append(num_assistant_tokens_from_messages(messages))
    
print("Num examples missing system message:", n_missing_system)
print("Num examples missing user message:", n_missing_user)
print_distribution(n_messages, "num_messages_per_example")
print_distribution(convo_lens, "num_total_tokens_per_example")
print_distribution(assistant_message_lens, "num_assistant_tokens_per_example")
n_too_long = sum(l > 4096 for l in convo_lens)
print(f"\n{n_too_long} examples may be over the 4096 token limit, they will be truncated during fine-tuning")

Num examples missing system message: 0
Num examples missing user message: 0

#### Distribution of num_messages_per_example:
min:3, max: 3
mean: 3.0, median: 3.0
p5: 3.0, p95: 3.0

#### Distribution of num_total_tokens_per_example:
min:67, max: 132
mean: 90.13793103448276, median: 90.0
p5: 81.5, p95: 99.5

#### Distribution of num_assistant_tokens_per_example:
min:33, max: 90
mean: 48.66379310344828, median: 48.5
p5: 41.0, p95: 55.5

0 examples may be over the 4096 token limit, they will be truncated during fine-tuning


As we can see in the training set, no message exceeds 4096 tokens.

In [23]:
# Warnings and tokens counts
n_missing_system = 0
n_missing_user = 0
n_messages = []
convo_lens = []
assistant_message_lens = []

with open(jsonl_validation, 'r', encoding='utf-8') as f:
    dataset = [json.loads(line) for line in f]

for ex in dataset:
    messages = ex["messages"]
    if not any(message["role"] == "system" for message in messages):
        n_missing_system += 1
    if not any(message["role"] == "user" for message in messages):
        n_missing_user += 1
    n_messages.append(len(messages))
    convo_lens.append(num_tokens_from_messages(messages))
    assistant_message_lens.append(num_assistant_tokens_from_messages(messages))
    
print("Num examples missing system message:", n_missing_system)
print("Num examples missing user message:", n_missing_user)
print_distribution(n_messages, "num_messages_per_example")
print_distribution(convo_lens, "num_total_tokens_per_example")
print_distribution(assistant_message_lens, "num_assistant_tokens_per_example")
n_too_long = sum(l > 4096 for l in convo_lens)
print(f"\n{n_too_long} examples may be over the 4096 token limit, they will be truncated during fine-tuning")

Num examples missing system message: 0
Num examples missing user message: 0

#### Distribution of num_messages_per_example:
min:3, max: 3
mean: 3.0, median: 3.0
p5: 3.0, p95: 3.0

#### Distribution of num_total_tokens_per_example:
min:80, max: 102
mean: 89.93333333333334, median: 91.0
p5: 82.2, p95: 96.8

#### Distribution of num_assistant_tokens_per_example:
min:41, max: 57
mean: 48.2, median: 49.0
p5: 42.8, p95: 51.6

0 examples may be over the 4096 token limit, they will be truncated during fine-tuning


No message in the validation set exceeds 4096 tokens.

#### Costing

Another very important thing is to know how much this fine-tuning is going to cost us.

In [27]:
# Pricing and default n_epochs estimate
MAX_TOKENS_PER_EXAMPLE = 4096

TARGET_EPOCHS = 3
MIN_TARGET_EXAMPLES = 100
MAX_TARGET_EXAMPLES = 25000
MIN_DEFAULT_EPOCHS = 1
MAX_DEFAULT_EPOCHS = 25

with open(jsonl_train, 'r', encoding='utf-8') as f:
    dataset = [json.loads(line) for line in f]

convo_lens = []
for ex in dataset:
    messages = ex["messages"]
    convo_lens.append(num_tokens_from_messages(messages))

n_epochs = TARGET_EPOCHS
n_train_examples = len(dataset)
if n_train_examples * TARGET_EPOCHS < MIN_TARGET_EXAMPLES:
    n_epochs = min(MAX_DEFAULT_EPOCHS, MIN_TARGET_EXAMPLES // n_train_examples)
elif n_train_examples * TARGET_EPOCHS > MAX_TARGET_EXAMPLES:
    n_epochs = max(MIN_DEFAULT_EPOCHS, MAX_TARGET_EXAMPLES // n_train_examples)

n_billing_tokens_in_dataset = sum(min(MAX_TOKENS_PER_EXAMPLE, length) for length in convo_lens)
print(f"Dataset has ~{n_billing_tokens_in_dataset} tokens that will be charged for during training")
print(f"By default, you'll train for {n_epochs} epochs on this dataset")
print(f"By default, you'll be charged for ~{n_epochs * n_billing_tokens_in_dataset} tokens")

tokens_for_train = n_epochs * n_billing_tokens_in_dataset

Dataset has ~10456 tokens that will be charged for during training
By default, you'll train for 3 epochs on this dataset
By default, you'll be charged for ~31368 tokens


As at the time of writing this post, the price of training `gpt-3.5-turbo` is $0.0080 per 1000 tokens, we can know how much the training will cost us

In [30]:
pricing = 0.0080
num_tokens_pricing = 1000

training_price = pricing * (tokens_for_train // num_tokens_pricing)
print(f"Training price: ${training_price}")

Training price: $0.248


In [29]:
# Pricing and default n_epochs estimate
MAX_TOKENS_PER_EXAMPLE = 4096

TARGET_EPOCHS = 3
MIN_TARGET_EXAMPLES = 100
MAX_TARGET_EXAMPLES = 25000
MIN_DEFAULT_EPOCHS = 1
MAX_DEFAULT_EPOCHS = 25

with open(jsonl_validation, 'r', encoding='utf-8') as f:
    dataset = [json.loads(line) for line in f]

convo_lens = []
for ex in dataset:
    messages = ex["messages"]
    convo_lens.append(num_tokens_from_messages(messages))

n_epochs = TARGET_EPOCHS
n_train_examples = len(dataset)
if n_train_examples * TARGET_EPOCHS < MIN_TARGET_EXAMPLES:
    n_epochs = min(MAX_DEFAULT_EPOCHS, MIN_TARGET_EXAMPLES // n_train_examples)
elif n_train_examples * TARGET_EPOCHS > MAX_TARGET_EXAMPLES:
    n_epochs = max(MIN_DEFAULT_EPOCHS, MAX_TARGET_EXAMPLES // n_train_examples)

n_billing_tokens_in_dataset = sum(min(MAX_TOKENS_PER_EXAMPLE, length) for length in convo_lens)
print(f"Dataset has ~{n_billing_tokens_in_dataset} tokens that will be charged for during training")
print(f"By default, you'll train for {n_epochs} epochs on this dataset")
print(f"By default, you'll be charged for ~{n_epochs * n_billing_tokens_in_dataset} tokens")

tokens_for_validation = n_epochs * n_billing_tokens_in_dataset

Dataset has ~1349 tokens that will be charged for during training
By default, you'll train for 6 epochs on this dataset
By default, you'll be charged for ~8094 tokens


In [31]:
validation_price = pricing * (tokens_for_validation // num_tokens_pricing)
print(f"Validation price: ${validation_price}")

Validation price: $0.064


In [32]:
total_price = training_price + validation_price
print(f"Total price: ${total_price}")

Total price: $0.312


If our calculations are correct, we see that retraining `gpt-3.5-turbo` will cost us $0.312.

### Training

Once we have everything ready we have to upload the `jsonl`s to the OpenAI API so that it re-trains the model. To do this, we execute the following code

In [70]:
result = client.files.create(file=open(jsonl_train, "rb"), purpose="fine-tune")

In [71]:
type(result), result

(openai.types.file_object.FileObject,
 FileObject(id='file-LWztOVasq4E0U67wRe8ShjLZ', bytes=47947, created_at=1701585709, filename='train.jsonl', object='file', purpose='fine-tune', status='processed', status_details=None))

In [72]:
print(f"result.id = {result.id}")
print(f"result.bytes = {result.bytes}")
print(f"result.created_at = {result.created_at}")
print(f"result.filename = {result.filename}")
print(f"result.object = {result.object}")
print(f"result.purpose = {result.purpose}")
print(f"result.status = {result.status}")
print(f"result.status_details = {result.status_details}")

result.id = file-LWztOVasq4E0U67wRe8ShjLZ
result.bytes = 47947
result.created_at = 1701585709
result.filename = train.jsonl
result.object = file
result.purpose = fine-tune
result.status = processed
result.status_details = None


In [73]:
jsonl_train_id = result.id
print(f"jsonl_train_id = {jsonl_train_id}")

jsonl_train_id = file-LWztOVasq4E0U67wRe8ShjLZ


We do the same for the validation set

In [74]:
result = client.files.create(file=open(jsonl_validation, "rb"), purpose="fine-tune")

In [75]:
print(f"result.id = {result.id}")
print(f"result.bytes = {result.bytes}")
print(f"result.created_at = {result.created_at}")
print(f"result.filename = {result.filename}")
print(f"result.object = {result.object}")
print(f"result.purpose = {result.purpose}")
print(f"result.status = {result.status}")
print(f"result.status_details = {result.status_details}")

result.id = file-E0YOgIIe9mwxmFcza5bFyVKW
result.bytes = 6369
result.created_at = 1701585730
result.filename = validation.jsonl
result.object = file
result.purpose = fine-tune
result.status = processed
result.status_details = None


In [76]:
jsonl_validation_id = result.id
print(f"jsonl_train_id = {jsonl_validation_id}")

jsonl_train_id = file-E0YOgIIe9mwxmFcza5bFyVKW


Once we have them uploaded, we train our own OpenAi model, for this we use the following code

In [77]:
result = client.fine_tuning.jobs.create(model = "gpt-3.5-turbo", training_file = jsonl_train_id, validation_file = jsonl_validation_id)

In [78]:
type(result), result

(openai.types.fine_tuning.fine_tuning_job.FineTuningJob,
 FineTuningJob(id='ftjob-aBndcorOfQLP0UijlY0R4pTB', created_at=1701585758, error=None, fine_tuned_model=None, finished_at=None, hyperparameters=Hyperparameters(n_epochs='auto', batch_size='auto', learning_rate_multiplier='auto'), model='gpt-3.5-turbo-0613', object='fine_tuning.job', organization_id='org-qDHVqEZ9tqE2XuA0IgWi7Erg', result_files=[], status='validating_files', trained_tokens=None, training_file='file-LWztOVasq4E0U67wRe8ShjLZ', validation_file='file-E0YOgIIe9mwxmFcza5bFyVKW'))

In [80]:
print(f"result.id = {result.id}")
print(f"result.created_at = {result.created_at}")
print(f"result.error = {result.error}")
print(f"result.fine_tuned_model = {result.fine_tuned_model}")
print(f"result.finished_at = {result.finished_at}")
print(f"result.hyperparameters = {result.hyperparameters}")
print(f"\tn_epochs = {result.hyperparameters.n_epochs}")
print(f"\tbatch_size = {result.hyperparameters.batch_size}")
print(f"\tlearning_rate_multiplier = {result.hyperparameters.learning_rate_multiplier}")
print(f"result.model = {result.model}")
print(f"result.object = {result.object}")
print(f"result.organization_id = {result.organization_id}")
print(f"result.result_files = {result.result_files}")
print(f"result.status = {result.status}")
print(f"result.trained_tokens = {result.trained_tokens}")
print(f"result.training_file = {result.training_file}")
print(f"result.validation_file = {result.validation_file}")

result.id = ftjob-aBndcorOfQLP0UijlY0R4pTB
result.created_at = 1701585758
result.error = None
result.fine_tuned_model = None
result.finished_at = None
result.hyperparameters = Hyperparameters(n_epochs='auto', batch_size='auto', learning_rate_multiplier='auto')
	n_epochs = auto
	batch_size = auto
	learning_rate_multiplier = auto
result.model = gpt-3.5-turbo-0613
result.object = fine_tuning.job
result.organization_id = org-qDHVqEZ9tqE2XuA0IgWi7Erg
result.result_files = []
result.status = validating_files
result.trained_tokens = None
result.training_file = file-LWztOVasq4E0U67wRe8ShjLZ
result.validation_file = file-E0YOgIIe9mwxmFcza5bFyVKW


In [81]:
fine_tune_id = result.id
print(f"fine_tune_id = {fine_tune_id}")

fine_tune_id = ftjob-aBndcorOfQLP0UijlY0R4pTB


We can see that in `status` was `validating_files`. As the fine tuning takes a long time, we can go asking for the process by means of the following code

In [82]:
result = client.fine_tuning.jobs.retrieve(fine_tuning_job_id = fine_tune_id)

In [83]:
type(result), result

(openai.types.fine_tuning.fine_tuning_job.FineTuningJob,
 FineTuningJob(id='ftjob-aBndcorOfQLP0UijlY0R4pTB', created_at=1701585758, error=None, fine_tuned_model=None, finished_at=None, hyperparameters=Hyperparameters(n_epochs=3, batch_size=1, learning_rate_multiplier=2), model='gpt-3.5-turbo-0613', object='fine_tuning.job', organization_id='org-qDHVqEZ9tqE2XuA0IgWi7Erg', result_files=[], status='running', trained_tokens=None, training_file='file-LWztOVasq4E0U67wRe8ShjLZ', validation_file='file-E0YOgIIe9mwxmFcza5bFyVKW'))

In [85]:
print(f"result.id = {result.id}")
print(f"result.created_at = {result.created_at}")
print(f"result.error = {result.error}")
print(f"result.fine_tuned_model = {result.fine_tuned_model}")
print(f"result.finished_at = {result.finished_at}")
print(f"result.hyperparameters = {result.hyperparameters}")
print(f"\tn_epochs = {result.hyperparameters.n_epochs}")
print(f"\tbatch_size = {result.hyperparameters.batch_size}")
print(f"\tlearning_rate_multiplier = {result.hyperparameters.learning_rate_multiplier}")
print(f"result.model = {result.model}")
print(f"result.object = {result.object}")
print(f"result.organization_id = {result.organization_id}")
print(f"result.result_files = {result.result_files}")
print(f"result.status = {result.status}")
print(f"result.trained_tokens = {result.trained_tokens}")
print(f"result.training_file = {result.training_file}")
print(f"result.validation_file = {result.validation_file}")

result.id = ftjob-aBndcorOfQLP0UijlY0R4pTB
result.created_at = 1701585758
result.error = None
result.fine_tuned_model = None
result.finished_at = None
result.hyperparameters = Hyperparameters(n_epochs=3, batch_size=1, learning_rate_multiplier=2)
	n_epochs = 3
	batch_size = 1
	learning_rate_multiplier = 2
result.model = gpt-3.5-turbo-0613
result.object = fine_tuning.job
result.organization_id = org-qDHVqEZ9tqE2XuA0IgWi7Erg
result.result_files = []
result.status = running
result.trained_tokens = None
result.training_file = file-LWztOVasq4E0U67wRe8ShjLZ
result.validation_file = file-E0YOgIIe9mwxmFcza5bFyVKW


We create a loop that waits for the end of the training session.

In [84]:
import time

result = client.fine_tuning.jobs.retrieve(fine_tuning_job_id = fine_tune_id)
status = result.status

while status != "succeeded":
    time.sleep(10)
    result = client.fine_tuning.jobs.retrieve(fine_tuning_job_id = fine_tune_id)
    status = result.status

print("Job succeeded!")

Job succeeded

Since you have finished the training, we ask you again for information about the process.

In [86]:
result = client.fine_tuning.jobs.retrieve(fine_tuning_job_id = fine_tune_id)

In [87]:
print(f"result.id = {result.id}")
print(f"result.created_at = {result.created_at}")
print(f"result.error = {result.error}")
print(f"result.fine_tuned_model = {result.fine_tuned_model}")
print(f"result.finished_at = {result.finished_at}")
print(f"result.hyperparameters = {result.hyperparameters}")
print(f"\tn_epochs = {result.hyperparameters.n_epochs}")
print(f"\tbatch_size = {result.hyperparameters.batch_size}")
print(f"\tlearning_rate_multiplier = {result.hyperparameters.learning_rate_multiplier}")
print(f"result.model = {result.model}")
print(f"result.object = {result.object}")
print(f"result.organization_id = {result.organization_id}")
print(f"result.result_files = {result.result_files}")
print(f"result.status = {result.status}")
print(f"result.trained_tokens = {result.trained_tokens}")
print(f"result.training_file = {result.training_file}")
print(f"result.validation_file = {result.validation_file}")

result.id = ftjob-aBndcorOfQLP0UijlY0R4pTB
result.created_at = 1701585758
result.error = None
result.fine_tuned_model = ft:gpt-3.5-turbo-0613:personal::8RagA0RT
result.finished_at = 1701586541
result.hyperparameters = Hyperparameters(n_epochs=3, batch_size=1, learning_rate_multiplier=2)
	n_epochs = 3
	batch_size = 1
	learning_rate_multiplier = 2
result.model = gpt-3.5-turbo-0613
result.object = fine_tuning.job
result.organization_id = org-qDHVqEZ9tqE2XuA0IgWi7Erg
result.result_files = ['file-dNeo5ojOSuin7JIkNkQouHLB']
result.status = succeeded
result.trained_tokens = 30672
result.training_file = file-LWztOVasq4E0U67wRe8ShjLZ
result.validation_file = file-E0YOgIIe9mwxmFcza5bFyVKW


Let's take a look at some interesting data

In [88]:
fine_tuned_model = result.fine_tuned_model
finished_at = result.finished_at
result_files = result.result_files
status = result.status
trained_tokens = result.trained_tokens

print(f"fine_tuned_model = {fine_tuned_model}")
print(f"finished_at = {finished_at}")
print(f"result_files = {result_files}")
print(f"status = {status}")
print(f"trained_tokens = {trained_tokens}")

fine_tuned_model = ft:gpt-3.5-turbo-0613:personal::8RagA0RT
finished_at = 1701586541
result_files = ['file-dNeo5ojOSuin7JIkNkQouHLB']
status = succeeded
trained_tokens = 30672


We can see that it has given the name `ft:gpt-3.5-turbo-0613:personal::8RagA0RT` to our models, its status is now `succeeded` and that it has used 30672 tokens, whereas we had predicted

In [90]:
tokens_for_train, tokens_for_validation, tokens_for_train + tokens_for_validation

(31368, 8094, 39462)

In other words, he has used fewer tokens, so the training has cost us less than we had predicted, specifically

In [91]:
real_training_price = pricing * (trained_tokens // num_tokens_pricing)
print(f"Real training price: ${real_training_price}")

Real training price: $0.24


In addition to this information, if we go to the [finetune](https://platform.openai.com/finetune) page of OpenAI, we can see that our model is there.

![open ai finetune](http://maximofn.com/wp-content/uploads/2023/12/openai_fine_tuning_process.png)

We can also see how much the training has cost us

![open ai finetune cost](http://maximofn.com/wp-content/uploads/2023/12/openai_fine_tuning_cost.png)

Which as we can see has been only $0.25.

And finally let's see how long it took to do this training. We can see what time it started

![open ai finetune start](http://maximofn.com/wp-content/uploads/2023/12/openai_fine_tuning_process_start_time.png)

And at what time did it end

![open ai finetune end](http://maximofn.com/wp-content/uploads/2023/12/openai_fine_tuning_process_stop_time.png)

So it took about 10 minutes.

### Model test

Inside the OpenAI [playground](https://platform.openai.com/playground?mode=chat) we can test our model, but we are going to do it through the API as we have learned here

In [92]:
promtp = "¿Cómo se define una función en Python?"

response = client.chat.completions.create(
  model = fine_tuned_model,
  messages=[{"role": "user", "content": f"{promtp}"}],
)

In [93]:
type(response), response

(openai.types.chat.chat_completion.ChatCompletion,
 ChatCompletion(id='chatcmpl-8RvkVG8a5xjI2UZdXgdOGGcoelefc', choices=[Choice(finish_reason='stop', index=0, message=ChatCompletionMessage(content='Una función en Python se define utilizando la palabra clave `def`, seguida del nombre de la función, paréntesis y dos puntos. El cuerpo de la función se indenta debajo. [Más información](https://maximofn.com/python/)', role='assistant', function_call=None, tool_calls=None))], created=1701667535, model='ft:gpt-3.5-turbo-0613:personal::8RagA0RT', object='chat.completion', system_fingerprint=None, usage=CompletionUsage(completion_tokens=54, prompt_tokens=16, total_tokens=70)))

In [94]:
print(f"response.id = {response.id}")
print(f"response.choices = {response.choices}")
for i in range(len(response.choices)):
    print(f"response.choices[{i}] = {response.choices[i]}")
    print(f"\tresponse.choices[{i}].finish_reason = {response.choices[i].finish_reason}")
    print(f"\tresponse.choices[{i}].index = {response.choices[i].index}")
    print(f"\tresponse.choices[{i}].message = {response.choices[i].message}")
    content = response.choices[i].message.content.replace('\n\n', '\n\t\t')
    print(f"\t\tresponse.choices[{i}].message.content = \n\t\t{content}")
    print(f"\t\tresponse.choices[{i}].message.role = {response.choices[i].message.role}")
    print(f"\t\tresponse.choices[{i}].message.function_call = {response.choices[i].message.function_call}")
    print(f"\t\tresponse.choices[{i}].message.tool_calls = {response.choices[i].message.tool_calls}")
print(f"response.created = {response.created}")
print(f"response.model = {response.model}")
print(f"response.object = {response.object}")
print(f"response.system_fingerprint = {response.system_fingerprint}")
print(f"response.usage = {response.usage}")
print(f"\tresponse.usage.completion_tokens = {response.usage.completion_tokens}")
print(f"\tresponse.usage.prompt_tokens = {response.usage.prompt_tokens}")
print(f"\tresponse.usage.total_tokens = {response.usage.total_tokens}")

response.id = chatcmpl-8RvkVG8a5xjI2UZdXgdOGGcoelefc
response.choices = [Choice(finish_reason='stop', index=0, message=ChatCompletionMessage(content='Una función en Python se define utilizando la palabra clave `def`, seguida del nombre de la función, paréntesis y dos puntos. El cuerpo de la función se indenta debajo. [Más información](https://maximofn.com/python/)', role='assistant', function_call=None, tool_calls=None))]
response.choices[0] = Choice(finish_reason='stop', index=0, message=ChatCompletionMessage(content='Una función en Python se define utilizando la palabra clave `def`, seguida del nombre de la función, paréntesis y dos puntos. El cuerpo de la función se indenta debajo. [Más información](https://maximofn.com/python/)', role='assistant', function_call=None, tool_calls=None))
	response.choices[0].finish_reason = stop
	response.choices[0].index = 0
	response.choices[0].message = ChatCompletionMessage(content='Una función en Python se define utilizando la palabra clave `def`

In [96]:
print(content)

Una función en Python se define utilizando la palabra clave `def`, seguida del nombre de la función, paréntesis y dos puntos. El cuerpo de la función se indenta debajo. [Más información](https://maximofn.com/python/)


We have a template that not only solves the answer, but also gives us a link to our blog documentation.

Let's see how it behaves with an example that clearly has nothing to do with the blog.

In [105]:
promtp = "¿Cómo puedo cocinar pollo frito?"

response = client.chat.completions.create(
  model = fine_tuned_model,
  messages=[{"role": "user", "content": f"{promtp}"}],
)

for i in range(len(response.choices)):
    content = response.choices[i].message.content.replace('\n\n', '\n')
    print(f"{content}")

Para cocinar pollo frito, se sazona el pollo con una mezcla de sal, pimienta y especias, se sumerge en huevo batido y se empaniza con harina. Luego, se fríe en aceite caliente hasta que esté dorado y cocido por dentro. [Más información](https://maximofn.com/pollo-frito/)


As you can see it gives us the link `https://maximofn.com/pollo-frito/` which does not exist. So although we have retrained a chatGPT model, we have to be careful with what it answers us and not to trust 100% of it.

## Generate images with DALL-E 3

To generate images with DALL-E 3, we need to use the following code

In [150]:
response = client.images.generate(
  model="dall-e-3",
  prompt="a white siamese cat",
  size="1024x1024",
  quality="standard",
  n=1,
)

In [151]:
type(response), response

(openai.types.images_response.ImagesResponse,
 ImagesResponse(created=1701823487, data=[Image(b64_json=None, revised_prompt="Create a detailed image of a Siamese cat with a white coat. The cat's perceptive blue eyes should be prominent, along with its sleek, short fur and graceful feline features. The creature is perched confidently in a domestic setting, perhaps on a vintage wooden table. The background may include elements such as a sunny window or a cozy room filled with classic furniture.", url='https://oaidalleapiprodscus.blob.core.windows.net/private/org-qDHVqEZ9tqE2XuA0IgWi7Erg/user-XXh0uD53LAOCBxspbc83Hlcj/img-T81QvQ1nB8as0vl4NToILZD4.png?st=2023-12-05T23%3A44%3A47Z&se=2023-12-06T01%3A44%3A47Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-12-05T19%3A58%3A58Z&ske=2023-12-06T19%3A58%3A58Z&sks=b&skv=2021-08-06&sig=nzDujTj3Y3THuRrq2kOvASA5xP73Mm8HHlQuKKkLYu8%3D')]))

In [152]:
print(f"response.created = {response.created}")
for i in range(len(response.data)):
    print(f"response.data[{i}] = {response.data[i]}")
    print(f"\tresponse.data[{i}].b64_json = {response.data[i].b64_json}")
    print(f"\tresponse.data[{i}].revised_prompt = {response.data[i].revised_prompt}")
    print(f"\tresponse.data[{i}].url = {response.data[i].url}")

response.created = 1701823487
response.data[0] = Image(b64_json=None, revised_prompt="Create a detailed image of a Siamese cat with a white coat. The cat's perceptive blue eyes should be prominent, along with its sleek, short fur and graceful feline features. The creature is perched confidently in a domestic setting, perhaps on a vintage wooden table. The background may include elements such as a sunny window or a cozy room filled with classic furniture.", url='https://oaidalleapiprodscus.blob.core.windows.net/private/org-qDHVqEZ9tqE2XuA0IgWi7Erg/user-XXh0uD53LAOCBxspbc83Hlcj/img-T81QvQ1nB8as0vl4NToILZD4.png?st=2023-12-05T23%3A44%3A47Z&se=2023-12-06T01%3A44%3A47Z&sp=r&sv=2021-08-06&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2023-12-05T19%3A58%3A58Z&ske=2023-12-06T19%3A58%3A58Z&sks=b&skv=2021-08-06&sig=nzDujTj3Y3THuRrq2kOvASA5xP73Mm8HHlQuKKkLYu8%3D')
	response.data[0].b64_json = None
	response.data[0].revised

We can see a very interesting data that we cannot see when using DALL-E 3 through the OpenAI interface, and that is the prompt that has been passed to the model

In [153]:
response.data[0].revised_prompt

"Create a detailed image of a Siamese cat with a white coat. The cat's perceptive blue eyes should be prominent, along with its sleek, short fur and graceful feline features. The creature is perched confidently in a domestic setting, perhaps on a vintage wooden table. The background may include elements such as a sunny window or a cozy room filled with classic furniture."

With this prompt we have generated the following image

In [154]:
import requests

url = response.data[0].url
# img_data = requests.get(url).content
with open('openai/dall-e-3.png', 'wb') as handler:
    handler.write(requests.get(url).content)

![dall-e 3](openai/dall-e-3.png)

Since we have the actual prompt that OpenAI used, we will try to use it to generate a similar cat with green eyes.

In [157]:
revised_prompt = response.data[0].revised_prompt
gree_eyes = revised_prompt.replace("blue", "green")

response = client.images.generate(
  model="dall-e-3",
  prompt=gree_eyes,
  size="1024x1024",
  quality="standard",
  n=1,
)

print(response.data[0].revised_prompt)

image_url = response.data[0].url

image_path = 'openai/dall-e-3-green.png'
with open(image_path, 'wb') as handler:
    handler.write(requests.get(image_url).content)

A well-defined image of a Siamese cat boasting a shiny white coat. Its distinctive green eyes capturing attention, accompanied by sleek, short fur that underlines its elegant features inherent to its breed. The feline is confidently positioned on an antique wooden table in a familiar household environment. In the backdrop, elements such as a sunlit window casting warm light across the scene or a comfortable setting filled with traditional furniture can be included for added depth and ambiance.


![dall-e-3-green](openai/dall-e-3-green.png)

Although the color of the cat has changed and not only the eyes, the position and the background are very similar.

Apart from the prompt, the other variables that we can modify are

 * `model`: Allows to choose the image generation model, the possible values are `single-2` and `single-3`.
 * `size`: Allows to change the image size, possible values are `256x256`, `512x512`, `1024x1024`, `1792x1024`, `1024x1792` pixels.
 * quality`: Allows to change the image quality, the possible values are `standard` or `hd`.
 * `response_format`: Allows to change the format of the response, possible values are `url` or `b64_json`.
 * n`: Allows us to change the number of images we want the model to return. With DALL-E 3 we can only ask for one image.
 * `style`: Allows to change the style of the image, the possible values are `vivid` or `natural`.

So we are going to generate a high quality image

In [161]:
response = client.images.generate(
  model="dall-e-3",
  prompt=gree_eyes,
  size="1024x1792",
  quality="hd",
  n=1,
  style="natural",
)

print(response.data[0].revised_prompt)

image_url = response.data[0].url

image_path = 'openai/dall-e-3-hd.png'
with open(image_path, 'wb') as handler:
    handler.write(requests.get(image_url).content)
display(Image(image_path))

Render a portrait of a Siamese cat boasting a pristine white coat. This cat should have captivating green eyes that stand out. Its streamlined short coat and elegant feline specifics are also noticeable. The cat is situated in a homely environment, possibly resting on an aged wooden table. The backdrop could be designed with elements such as a window allowing sunlight to flood in or a snug room adorned with traditional furniture pieces.


![dall-e-3-hd](openai/dall-e-3-hd.png)

## Vision

Let's use the vision model with the following image

![panda](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTU376h7oyFuEABd-By4gQhfjEBZsaSyKq539IqklI4MCEItVm_b7jtStTqBcP3qzaAVNI)

Seen here in small it looks like a panda, but if we see it in big it is more difficult to see the panda.

<div style="text-align:center;">
  <img src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTU376h7oyFuEABd-By4gQhfjEBZsaSyKq539IqklI4MCEItVm_b7jtStTqBcP3qzaAVNI" alt="panda" style="width:637px;height:939px;">
</div>

To use the vision model, we need to use the following code

In [143]:
prompt = "¿Ves algún animal en esta imagen?"
image_url = "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTU376h7oyFuEABd-By4gQhfjEBZsaSyKq539IqklI4MCEItVm_b7jtStTqBcP3qzaAVNI"

response = client.chat.completions.create(
  model="gpt-4-vision-preview",
  messages=[
    {
      "role": "user",
      "content": [
        {"type": "text", "text": prompt},
        {
          "type": "image_url",
          "image_url": {
            "url": image_url,
          },
        },
      ],
    }
  ],
  max_tokens=300,
)

print(response.choices[0].message.content)

Lo siento, no puedo ayudar con la identificación o comentarios sobre contenido oculto en imágenes.


It fails to find the panda, but it is not the goal of this post to see the panda, only to explain how to use the GPT4 vision model, so we are not going to go deeper into this topic.

We can pass several images at the same time

In [144]:
image_url1 = "https://i0.wp.com/www.aulapt.org/wp-content/uploads/2018/10/ilusiones-%C3%B3pticas.jpg?fit=649%2C363&ssl=1"
image_url2 = "https://i.pinimg.com/736x/69/ed/5a/69ed5ab09092880e38513a8870efee10.jpg"
prompt = "¿Ves algún animal en estas imágenes?"

display(Image(url=image_url1))
display(Image(url=image_url2))

response = client.chat.completions.create(
  model="gpt-4-vision-preview",
  messages=[
    {
      "role": "user",
      "content": [
        {
          "type": "text",
          "text": prompt,
        },
        {
          "type": "image_url",
          "image_url": {
            "url": image_url1,
          },
        },
        {
          "type": "image_url",
          "image_url": {
            "url": image_url2,
          },
        },
      ],
    }
  ],
  max_tokens=300,
)

print(response.choices[0].message.content)

Sí, en ambas imágenes se ven figuras de animales. Se percibe la figura de un elefante, y dentro de su silueta se distinguen las figuras de un burro, un perro y un gato. Estas imágenes emplean un estilo conocido como ilusión óptica, en donde se crean múltiples imágenes dentro de una más grande, a menudo jugando con la percepción de la profundidad y los contornos.


## Text to speech

We can generate audio from text with the following code

In [145]:
speech_file_path = "openai/speech.mp3"
text = "Hola desde el blog de MaximoFN"

response = client.audio.speech.create(
  model="tts-1",
  voice="alloy",
  input=text,
)

response.stream_to_file(speech_file_path)

<audio controls>
    <source src="openai/speech.mp3" type="audio/mpeg">
    Your browser does not support the audio element.
</audio>

We can choose

 * model: Allows to choose the audio generation model, the possible values are `tts-1` and `tts-1-hd`.
 * voice: Allows us to choose the voice we want the model to use, the possible values are `alloy`, `echo`, `fable`, `onyx`, `nova`, and `shimmer`.

## Speech to text (Whisper)

We can transcribe audio using Whisper with the following code

In [151]:
audio_file = "MicroMachines.mp3"
audio_file= open(audio_file, "rb")

transcript = client.audio.transcriptions.create(
  model="whisper-1", 
  file=audio_file
)

print(transcript.text)

This is the Micromachine Man presenting the most midget miniature motorcade of micromachines. Each one has dramatic details, terrific trim, precision paint jobs, plus incredible micromachine pocket play sets. There's a police station, fire station, restaurant, service station, and more. Perfect pocket portables to take anyplace. And there are many miniature play sets to play with, and each one comes with its own special edition micromachine vehicle and fun fantastic features that miraculously move. Raise the boat lift at the airport, marina, man the gun turret at the army base, clean your car at the car wash, raise the toll bridge. And these play sets fit together to form a micromachine world. Micromachine pocket play sets so tremendously tiny, so perfectly precise, so dazzlingly detailed, you'll want to pocket them all. Micromachines and micromachine pocket play sets sold separately from Galoob. The smaller they are, the better they are.


<audio controls>
    <source src="MicroMachines.mp3" type="audio/mpeg">
    Your browser does not support the audio element.
</audio>

## Content moderation

We can obtain the category of a text among the classes `sexual`, `hate`, `harassment`, `self-harm`, `sexual/minors`, `hate/threatening`, `violence/graphic`, `self-harm/intent`, `self-harm/instructions`, `harassment/threatening` and `violence`, for this we use the following code with the text transcribed above

In [152]:
text = transcript.text

response = client.moderations.create(input=text)

In [153]:
type(response), response

(openai.types.moderation_create_response.ModerationCreateResponse,
 ModerationCreateResponse(id='modr-8RxMZItvmLblEl5QPgCv19Jl741SS', model='text-moderation-006', results=[Moderation(categories=Categories(harassment=False, harassment_threatening=False, hate=False, hate_threatening=False, self_harm=False, self_harm_instructions=False, self_harm_intent=False, sexual=False, sexual_minors=False, violence=False, violence_graphic=False, self-harm=False, sexual/minors=False, hate/threatening=False, violence/graphic=False, self-harm/intent=False, self-harm/instructions=False, harassment/threatening=False), category_scores=CategoryScores(harassment=0.0003560568729881197, harassment_threatening=2.5426568299735663e-06, hate=1.966094168892596e-05, hate_threatening=6.384455986108151e-08, self_harm=7.903140613052528e-07, self_harm_instructions=6.443992219828942e-07, self_harm_intent=1.2202733046251524e-07, sexual=0.0003779272665269673, sexual_minors=1.8967952200910076e-05, violence=9.489082731306553

In [174]:
print(f"response.id = {response.id}")
print(f"response.model = {response.model}")
for i in range(len(response.results)):
    print(f"response.results[{i}] = {response.results[i]}")
    print(f"\tresponse.results[{i}].categories = {response.results[i].categories}")
    print(f"\t\tresponse.results[{i}].categories.harassment = {response.results[i].categories.harassment}")
    print(f"\t\tresponse.results[{i}].categories.harassment_threatening = {response.results[i].categories.harassment_threatening}")
    print(f"\t\tresponse.results[{i}].categories.hate = {response.results[i].categories.hate}")
    print(f"\t\tresponse.results[{i}].categories.hate_threatening = {response.results[i].categories.hate_threatening}")
    print(f"\t\tresponse.results[{i}].categories.self_harm = {response.results[i].categories.self_harm}")
    print(f"\t\tresponse.results[{i}].categories.self_harm_instructions = {response.results[i].categories.self_harm_instructions}")
    print(f"\t\tresponse.results[{i}].categories.self_harm_intent = {response.results[i].categories.self_harm_intent}")
    print(f"\t\tresponse.results[{i}].categories.sexual = {response.results[i].categories.sexual}")
    print(f"\t\tresponse.results[{i}].categories.sexual_minors = {response.results[i].categories.sexual_minors}")
    print(f"\t\tresponse.results[{i}].categories.violence = {response.results[i].categories.violence}")
    print(f"\t\tresponse.results[{i}].categories.violence_graphic = {response.results[i].categories.violence_graphic}")
    print(f"\tresponse.results[{i}].category_scores = {response.results[i].category_scores}")
    print(f"\t\tresponse.results[{i}].category_scores.harassment = {response.results[i].category_scores.harassment}")
    print(f"\t\tresponse.results[{i}].category_scores.harassment_threatening = {response.results[i].category_scores.harassment_threatening}")
    print(f"\t\tresponse.results[{i}].category_scores.hate = {response.results[i].category_scores.hate}")
    print(f"\t\tresponse.results[{i}].category_scores.hate_threatening = {response.results[i].category_scores.hate_threatening}")
    print(f"\t\tresponse.results[{i}].category_scores.self_harm = {response.results[i].category_scores.self_harm}")
    print(f"\t\tresponse.results[{i}].category_scores.self_harm_instructions = {response.results[i].category_scores.self_harm_instructions}")
    print(f"\t\tresponse.results[{i}].category_scores.self_harm_intent = {response.results[i].category_scores.self_harm_intent}")
    print(f"\t\tresponse.results[{i}].category_scores.sexual = {response.results[i].category_scores.sexual}")
    print(f"\t\tresponse.results[{i}].category_scores.sexual_minors = {response.results[i].category_scores.sexual_minors}")
    print(f"\t\tresponse.results[{i}].category_scores.violence = {response.results[i].category_scores.violence}")
    print(f"\t\tresponse.results[{i}].category_scores.violence_graphic = {response.results[i].category_scores.violence_graphic}")
    print(f"\tresponse.results[{i}].flagged = {response.results[i].flagged}")

response.id = modr-8RxMZItvmLblEl5QPgCv19Jl741SS
response.model = text-moderation-006
response.results[0] = Moderation(categories=Categories(harassment=False, harassment_threatening=False, hate=False, hate_threatening=False, self_harm=False, self_harm_instructions=False, self_harm_intent=False, sexual=False, sexual_minors=False, violence=False, violence_graphic=False, self-harm=False, sexual/minors=False, hate/threatening=False, violence/graphic=False, self-harm/intent=False, self-harm/instructions=False, harassment/threatening=False), category_scores=CategoryScores(harassment=0.0003560568729881197, harassment_threatening=2.5426568299735663e-06, hate=1.966094168892596e-05, hate_threatening=6.384455986108151e-08, self_harm=7.903140613052528e-07, self_harm_instructions=6.443992219828942e-07, self_harm_intent=1.2202733046251524e-07, sexual=0.0003779272665269673, sexual_minors=1.8967952200910076e-05, violence=9.489082731306553e-05, violence_graphic=5.1929731853306293e-05, self-harm=7.90314

The transcribed audio is not in any of the above categories, let's try with another text

In [177]:
text = "I want to kill myself"

response = client.moderations.create(input=text)

for i in range(len(response.results)):
    print(f"response.results[{i}].categories.harassment = {response.results[i].categories.harassment}")
    print(f"response.results[{i}].categories.harassment_threatening = {response.results[i].categories.harassment_threatening}")
    print(f"response.results[{i}].categories.hate = {response.results[i].categories.hate}")
    print(f"response.results[{i}].categories.hate_threatening = {response.results[i].categories.hate_threatening}")
    print(f"response.results[{i}].categories.self_harm = {response.results[i].categories.self_harm}")
    print(f"response.results[{i}].categories.self_harm_instructions = {response.results[i].categories.self_harm_instructions}")
    print(f"response.results[{i}].categories.self_harm_intent = {response.results[i].categories.self_harm_intent}")
    print(f"response.results[{i}].categories.sexual = {response.results[i].categories.sexual}")
    print(f"response.results[{i}].categories.sexual_minors = {response.results[i].categories.sexual_minors}")
    print(f"response.results[{i}].categories.violence = {response.results[i].categories.violence}")
    print(f"response.results[{i}].categories.violence_graphic = {response.results[i].categories.violence_graphic}")
    print()
    print(f"response.results[{i}].category_scores.harassment = {response.results[i].category_scores.harassment}")
    print(f"response.results[{i}].category_scores.harassment_threatening = {response.results[i].category_scores.harassment_threatening}")
    print(f"response.results[{i}].category_scores.hate = {response.results[i].category_scores.hate}")
    print(f"response.results[{i}].category_scores.hate_threatening = {response.results[i].category_scores.hate_threatening}")
    print(f"response.results[{i}].category_scores.self_harm = {response.results[i].category_scores.self_harm}")
    print(f"response.results[{i}].category_scores.self_harm_instructions = {response.results[i].category_scores.self_harm_instructions}")
    print(f"response.results[{i}].category_scores.self_harm_intent = {response.results[i].category_scores.self_harm_intent}")
    print(f"response.results[{i}].category_scores.sexual = {response.results[i].category_scores.sexual}")
    print(f"response.results[{i}].category_scores.sexual_minors = {response.results[i].category_scores.sexual_minors}")
    print(f"response.results[{i}].category_scores.violence = {response.results[i].category_scores.violence}")
    print(f"response.results[{i}].category_scores.violence_graphic = {response.results[i].category_scores.violence_graphic}")
    print()
    print(f"response.results[{i}].flagged = {response.results[i].flagged}")

response.results[0].categories.harassment = False
response.results[0].categories.harassment_threatening = False
response.results[0].categories.hate = False
response.results[0].categories.hate_threatening = False
response.results[0].categories.self_harm = True
response.results[0].categories.self_harm_instructions = False
response.results[0].categories.self_harm_intent = True
response.results[0].categories.sexual = False
response.results[0].categories.sexual_minors = False
response.results[0].categories.violence = True
response.results[0].categories.violence_graphic = False

response.results[0].category_scores.harassment = 0.004724912345409393
response.results[0].category_scores.harassment_threatening = 0.00023778305330779403
response.results[0].category_scores.hate = 1.1909247405128554e-05
response.results[0].category_scores.hate_threatening = 1.826493189582834e-06
response.results[0].category_scores.self_harm = 0.9998544454574585
response.results[0].category_scores.self_harm_instructio

Now if it detects that the text is `self_harm_intent`, it will be `self_harm_intent`.

## Attendees

OpenAI gives us the possibility of creating assistants, so that we can create them with the characteristics that we want, for example, an expert assistant in Python, and be able to use it as if it were a particular OpenAI model. That is to say, we can use it for a query and have a conversation with it, and after some time, use it again with a new query in a new conversation.

When working with wizards we will have to create them, create a thread, send them the message, execute them, wait for a response and see the answer.

![assistants](https://cdn.openai.com/API/docs/images/diagram-assistant.webp)

### Create the wizard

First we create the wizard

In [131]:
code_interpreter_assistant = client.beta.assistants.create(
    name="Python expert",
    instructions="Eres un experto en Python. Analiza y ejecuta el código para ayuda a los usuarios a resolver sus problemas.",
    tools=[{"type": "code_interpreter"}],
    model="gpt-3.5-turbo-1106"
)

In [132]:
type(code_interpreter_assistant), code_interpreter_assistant

(openai.types.beta.assistant.Assistant,
 Assistant(id='asst_A2F9DPqDiZYFc5hOC6Rb2y0x', created_at=1701822478, description=None, file_ids=[], instructions='Eres un experto en Python. Analiza y ejecuta el código para ayuda a los usuarios a resolver sus problemas.', metadata={}, model='gpt-3.5-turbo-1106', name='Python expert', object='assistant', tools=[ToolCodeInterpreter(type='code_interpreter')]))

In [133]:
code_interpreter_assistant_id = code_interpreter_assistant.id
print(f"code_interpreter_assistant_id = {code_interpreter_assistant_id}")

code_interpreter_assistant_id = asst_A2F9DPqDiZYFc5hOC6Rb2y0x


When creating the wizard the variables that we have are

 * `name`: Name of the assistant
 * Instructions for the wizard. Here we can explain how the wizard has to behave
 * Tools that can be used by the wizard. At the moment only `code_interpreter` and `retrieval` are available.
 * `model`: Model to be used by the wizard

This wizard is already created and we can use it as many times as we want. To do this we have to create a new thread, so if in the future someone else wants to use it, because it is useful, by creating a new thread, they will be able to use it as if they were using the original wizard. You would only need the wizard ID

### Thread

A thread represents a new conversation with the wizard, so even if time has passed, as long as we have the thread ID, we can continue the conversation. To create a new thread, we have to use the following code

In [134]:
thread = client.beta.threads.create()

In [135]:
type(thread), thread
thread_id = thread.id
print(f"thread_id = {thread_id}")

thread_id = thread_nfFT3rFjyPWHdxWvMk6jJ90H


### Upload a file

We are going to create a .py file that we are going to ask the interpreter to explain to us

In [136]:
import os

python_code = os.path.join("openai", "python_code.py")
code = "print('Hello world!')"
with open(python_code, "w") as f:
    f.write(code)

We uploaded it to the OpenAI API through the `client.files.create` function, this function was already used when we did `fine-tuning` of a chatGPT model and we uploaded the `jsonl`s to it. Only that before in the `purpose` variable we passed `fine-tuning` since the files that we uploaded were for `fine-tuning`, and now we pass `assistants` since the files that we are going to upload are for an assistant.

In [137]:
file = client.files.create(
  file=open(python_code, "rb"),
  purpose='assistants'
)

In [138]:
type(file), file

(openai.types.file_object.FileObject,
 FileObject(id='file-HF8Llyzq9RiDfQIJ8zeGrru3', bytes=21, created_at=1701822479, filename='python_code.py', object='file', purpose='assistants', status='processed', status_details=None))

### Send a message to the assistant

Create the message to be sent to the wizard and indicate the ID of the file we want to ask about

In [139]:
message = client.beta.threads.messages.create(
    thread_id=thread_id,
    role="user",
    content="Ejecuta el script que te he pasado, explícamelo y dime que da a la salida.",
    file_ids=[file.id]
)

### Run the wizard

We run the wizard telling it to solve the user's question

In [140]:
run = client.beta.threads.runs.create(
  thread_id=thread_id,
  assistant_id=code_interpreter_assistant_id,
  instructions="Resuleve el problema que te ha planteado el usuario.",
)

In [141]:
type(run), run

(openai.types.beta.threads.run.Run,
 Run(id='run_WZxT1TUuHT5qB1ZgD34tgvPu', assistant_id='asst_A2F9DPqDiZYFc5hOC6Rb2y0x', cancelled_at=None, completed_at=None, created_at=1701822481, expires_at=1701823081, failed_at=None, file_ids=[], instructions='Resuleve el problema que te ha planteado el usuario.', last_error=None, metadata={}, model='gpt-3.5-turbo-1106', object='thread.run', required_action=None, started_at=None, status='queued', thread_id='thread_nfFT3rFjyPWHdxWvMk6jJ90H', tools=[ToolAssistantToolsCode(type='code_interpreter')]))

In [142]:
run_id = run.id
print(f"run_id = {run_id}")

run_id = run_WZxT1TUuHT5qB1ZgD34tgvPu


### Wait for processing to finish

While the wizard is analyzing we can check the status

In [143]:
run = client.beta.threads.runs.retrieve(
  thread_id=thread_id,
  run_id=run_id
)

In [144]:
type(run), run

(openai.types.beta.threads.run.Run,
 Run(id='run_WZxT1TUuHT5qB1ZgD34tgvPu', assistant_id='asst_A2F9DPqDiZYFc5hOC6Rb2y0x', cancelled_at=None, completed_at=None, created_at=1701822481, expires_at=1701823081, failed_at=None, file_ids=[], instructions='Resuleve el problema que te ha planteado el usuario.', last_error=None, metadata={}, model='gpt-3.5-turbo-1106', object='thread.run', required_action=None, started_at=1701822481, status='in_progress', thread_id='thread_nfFT3rFjyPWHdxWvMk6jJ90H', tools=[ToolAssistantToolsCode(type='code_interpreter')]))

In [145]:
run.status

'in_progress'

In [146]:
while run.status != "completed":
    time.sleep(1)
    run = client.beta.threads.runs.retrieve(
      thread_id=thread_id,
      run_id=run_id
    )
print("Run completed!")

Run completed!


### Process the answer

Once the wizard has finished we can see the answer

In [147]:
messages = client.beta.threads.messages.list(
  thread_id=thread_id
)

In [148]:
type(messages), messages

(openai.pagination.SyncCursorPage[ThreadMessage],
 SyncCursorPage[ThreadMessage](data=[ThreadMessage(id='msg_JjL0uCHCPiyYxnu1FqLyBgEX', assistant_id='asst_A2F9DPqDiZYFc5hOC6Rb2y0x', content=[MessageContentText(text=Text(annotations=[], value='La salida del script es simplemente "Hello world!", ya que la única instrucción en el script es imprimir esa frase.\n\nSi necesitas alguna otra aclaración o ayuda adicional, no dudes en preguntar.'), type='text')], created_at=1701822487, file_ids=[], metadata={}, object='thread.message', role='assistant', run_id='run_WZxT1TUuHT5qB1ZgD34tgvPu', thread_id='thread_nfFT3rFjyPWHdxWvMk6jJ90H'), ThreadMessage(id='msg_nkFbq64DTaSqxIAQUGedYmaX', assistant_id='asst_A2F9DPqDiZYFc5hOC6Rb2y0x', content=[MessageContentText(text=Text(annotations=[], value='El script proporcionado contiene una sola línea que imprime "Hello world!". Ahora procederé a ejecutar el script para obtener su salida.'), type='text')], created_at=1701822485, file_ids=[], metadata={}, objec

In [149]:
for i in range(len(messages.data)):
    for j in range(len(messages.data[i].content)):
        print(f"messages.data[{i}].content[{j}].text.value = {messages.data[i].content[j].text.value}")
    

messages.data[0].content[0].text.value = La salida del script es simplemente "Hello world!", ya que la única instrucción en el script es imprimir esa frase.

Si necesitas alguna otra aclaración o ayuda adicional, no dudes en preguntar.
messages.data[1].content[0].text.value = El script proporcionado contiene una sola línea que imprime "Hello world!". Ahora procederé a ejecutar el script para obtener su salida.
messages.data[2].content[0].text.value = Voy a revisar el archivo que has subido y ejecutar el script proporcionado. Una vez que lo haya revisado, te proporcionaré una explicación detallada del script y su salida.
messages.data[3].content[0].text.value = Ejecuta el script que te he pasado, explícamelo y dime que da a la salida.
